README
Kafka package version 1.08
==============================

The Kafka package is a set of Perl modules which provides a
simple and consistent application programming interface (API) to the
Apache Kafka 0.9+, a high-throughput distributed messaging system.
The main focus of the package is to provide classes
and functions that allow you to write Apache Kafka clients.


The Kafka package was written, tested, and found working on
recent Linux distributions.

INSTALLATION

To install this package type the following:

   perl Build.PL
   ./Build
   ./Build test
   ./Build install

DEPENDENCIES

In order to install and use this package you will need Perl version
5.010 or higher. Some modules within this package depend on other
packages that are distributed separately from Perl. We recommend that
you have the following packages installed before you install
Kafka:

   Authen::SCRAM::Client
   Carp
   Compress::Snappy
   Const::Fast
   Data::Compare
   Data::HexDump::Range
   Data::Validate::Domain
   Data::Validate::IP
   Exception::Class
   Exporter
   Gzip::Faster
   IO::Select
   List::Util
   Params::Util
   Scalar::Util
   Scalar::Util::Numeric
   Socket
   String::CRC32
   Sys::SigAction
   Time::HiRes
   Try::Tiny

The Kafka package has the following optional dependencies:

   Capture::Tiny
   Carp
   Clone
   Config::IniFiles
   Const::Fast
   Cwd
   Data::Dumper
   Exporter
   File::HomeDir
   File::Path
   File::Spec
   File::Spec::Functions
   IO::File
   IO::Socket::INET
   IO::Socket::IP
   List::Util
   Net::EmptyPort
   Params::Util
   Proc::Daemon
   Scalar::Util
   Scalar::Util::Numeric
   Socket
   Sub::Install
   Sys::SigAction
   Test::Deep
   Test::Exception
   Test::Fixme
   Test::More
   Test::NoWarnings
   Test::TCP
   Time::HiRes
   Try::Tiny

If the optional modules are missing, some "prereq" tests are skipped.

The installation of the missing dependencies can either be accomplished
through your OS package manager or through CPAN (or downloading the source
for all dependencies and compiling them manually).

TESTING

The following environment variable changes the way Kafka Benchmark test behave:

   # The root directory of the installation Kafka 0.9+
   KAFKA_BASE_DIR

SIMPLE BENCHMARK

Use tools/benchmark_consumer.pl and tools/benchmark_producer.pl to run a simple
benchmark.
Run one or more consumer or producer instances using tools/benchmark_consumer.pl or
tools/benchmark_producer.pl .

DOCUMENTATION

Every Kafka package module has a man page. A simple description of each module is
included below in the SEE ALSO section.

SEE ALSO

The basic operation of the Kafka package modules:

   Kafka             - constants and messages used by the Kafka package.
   Kafka::Connection - API to connect to a kafka cluster.
   Kafka::Producer   - interface to the 'producer' client.
   Kafka::Consumer   - interface to the 'consumer' client.
   Kafka::Message    - interface to the Kafka message properties.
   Kafka::Int64      - functions to work with 64 bit elements of the
                       protocol on 32 bit systems.
   Kafka::Protocol   - functions to process messages in the
                       Apache Kafka's Protocol.
   Kafka::IO         - low-level communications with the Apache Kafka server.
   Kafka::Exceptions - module designated to handle Kafka exceptions.
   Kafka::Internals  - Internal constants and functions used
                       by several packaged modules.

COPYRIGHT AND LICENCE

Copyright (C) 2012-2017 by TrackingSoft LLC.

This package is free software; you can redistribute it and/or modify it under
the same terms as Perl itself. See perlartistic at
http://dev.perl.org/licenses/artistic.html .

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
or FITNESS FOR A PARTICULAR PURPOSE.


Changes
Revision history for Kafka package

1.08
    Add support SASL authentication support

    Use Dist::Zilla static_install

    Migrate to milla

1.07 Sun September 17 2017

    Added support for v1 Messages at Producer side, which allows
    to produce messages with CreateTime timestamp. 

    Added support for LZ4 compression.

    BUGFIXES:
    - Fix Snappy compression to be compatible with xerial's
      snappy-java multiframes.

1.06 Fri July 28 2017

    Important fix for offset commit/fetch bug

    In Kafka versions > 0.8.2 offset commits/fetch are stored on
    Kafka itself, not zookeeper (that's version 1 of protocol
    instead of version 0).

    The code before that commit was claiming that it supported
    version 0 for commiting offsets, and version 1 for fetching.
    As a result, offsets were stored on zookeeper, but fetched
    from Kafka, where they were never found.

    This fix provides support for version 1 when storing and
    fetching offsets. It required implementing the API call for
    finding a group coordinator as well.

1.05 Tue June 20 2017

    BUGFIXES:
    - Fix message offsets for messages in compressed MessageSet (Damien Krotkine).
    - Do not use $_ in IO::connect (Petr Plavjanik).

1.04 Wed June 07 2017

    CHANGES:
    - Added offset commit/fetch API (thanks to Greg Franklin and Damien Krotkine).

    BUGFIXES:
    - Fix key-value arguments validation.
    - Fix MessageSet incomplete data check.

1.03 Mon May  22 2017

    BUGFIXES:
    - Prevent exceptions with Kafka 0.9 (where API versioning is not supported).

1.02 Fri May  12 2017

    Special thanks to Damien Krotkine for API versioning implementation.

    CHANGES:
    - Added API versioning support with auto-detection.
    - Timestamps in messages.
    - Improvements in offset methods: to be able to query offsets in an
      easier, simpler and more complete way (including support for
      version 0 and version 1 of message formats)
    - Support for array of keys in Kafka::Producer->send with array of messages.
    - Use Gzip::Faster to improve compression speed.

    BUGFIXES:
    - Improved timeouts handling for big payloads in Kafka::IO.

1.001013 Fri Feb  24 2017
    CHANGES:
    - Removed support for custom CorrelationID in constructor arguments.
    - Use seconds for Consumer.MaxWaitTime to make it similar to Producer.Timeout.
    - Removed deprecated constants $RECEIVE_MAX_ATTEMPTS, $RECEIVE_MAX_RETRIES, $SEND_MAX_RETRIES.
    - Replaced Kafka::Connection->is_server_alive, Kafka::IO->is_alive with private methods (to be
      used for internal tests only).

    BUGFIXES:
    - Fixed incorrect timeout values in requests.
    - Improved socket handling in Kafka::IO.

1.001012 Thu Nov  17 2016
    BUGFIXES:
    - Do not use $Kafka::SEND_MAX_ATTEMPTS in Kafka::Producer->send request
      to prevent duplicates.

1.001011 Wed Nov  09 2016
    CHANGES:
    - Decrease tcp connections creation.
    BUGFIXES:
    - The dependency declaration corrected.

1.001001 Tue Jun  07 2016
    CHANGES:
    - Version schema changed to reflect supported Kafka version.
      N.XXYYZZ where XXYY - major Kafka version N and ZZ - major and minor
      versions of Perl module.
    - Adapted for kafka 0.10.0.0 .
    - Added new Apache Kafka server error codes.
    - $MESSAGE_SIZE_OVERHEAD increased to 34.

0.9012 Thu May  26 2016
    CHANGES:
    - Test improvements.
    BUGFIXES:
    - Data::Dumper 'Sparseseen' method not used now.

0.9011 Tue May  24 2016
    BUGFIXES:
    - Test::More required version corrected.

0.9010 Thu Apr  28 2016
    CHANGES:
    - Adapted for kafka 0.9.0.1 .
    BUGFIXES:
    - Tests corrected.
    - Require perl 5.10 core modules version.

0.9001 Tri Feb  05 2016
    BUGFIXES:
    - Tests corrected.

0.9000 Thu Feb  04 2016
    CHANGES:
    - Adapted for kafka 0.9 .
    - Updated error codes to conform with Kafka documentation.
    - Kafka::Producer->new does not support values of RequiredAcks > 1 anymore.
    - Added 'ip_version' argument for Kafka::Connection->new and Kafka::IO->new to
      force host name resolving (or IP address interpretation) in IPv4 or IPv6.
    - Added error codes $ERROR_INCOMPATIBLE_HOST_IP_VERSION .
    BUGFIXES:
    - Incorrect osvers comparison for NetBSD.
    - Remove unnecessary dependency on List::MoreUtils
      (thanks to Dan Book).

0.8013 Tue Nov  24 2015
    BUGFIXES:
    - Packing timeval for NetBSD 6.0 and later.
    - Use inet_aton instead inet_pton for IPv4 addresses.
    - t/03_io_ipv6.t fixed to work with perl earlier than 5.014 .

0.8012 Mon Nov  09 2015
    CHANGES:
    - Support IPv6. IPv6 address in standard form.
    BUGFIXES:
    - Fix MANIFEST.
    - Error analysis is improved in Kafka::Connection .

0.8011 Mon Aug  18 2015
    CHANGES:
    - Kafka::Connection connection is closed only when $connection->close method is called
      explicitly.
    - Kafka::IO object destruction do not close socket.
    BUGFIXES:
    - Added Apache Kafka server error codes:
      $ERROR_LOAD_IN_PROGRESS_CODE,
      $ERROR_CONSUMER_COORDINATOR_NOT_AVAILABLE_CODE,
      $ERROR_NOT_COORDINATOR_FOR_CONSUMER_CODE
    - $ERROR_REPLICA_NOT_AVAILABLE
      error ignored according to Apache Kafka documentation
      (reported by Shabnam Shamaei).
    - Kafka::IO->is_alive method correctly determines server availability.

0.8010 Fri Feb  06 2015
    CHANGES:
    - $RECEIVE_MAX_RETRIES, $SEND_MAX_RETRIES renamed
      to $RECEIVE_MAX_ATTEMPTS, $SEND_MAX_ATTEMPTS .
    - Method Kafka::Connection->is_server_alive renamed to is_server_connected .
    - New Kafka::Connection->is_server_alive method properly checks
      for availability of Kafka server.
    - Added method Kafka::Connection->get_metadata .
    - Added method Kafka::Connection->exists_topic_partition .
    - $Kafka::MIN_BYTES_RESPOND_HAS_DATA changed to 10.
    - Testing supports Kafka version 0.8.2 .
    BUGFIXES:
    - Partititon cannot be negative.

Makefile.PL
#!/usr/bin/perl -w

use strict;
use warnings;
use 5.010;

if ( $^O eq 'MSWin32' || $^O eq 'solaris' ||  $^O eq 'cygwin' ) {
    print STDERR "OS unsupported: $^O\n";
    exit( 255 );
}

use ExtUtils::MakeMaker qw(
    prompt
);
use File::Spec::Functions qw(
    catdir
);

my %config = (
    NAME                => 'Kafka',
    VERSION_FROM        => 'lib/Kafka.pm',  # finds $VERSION
    MIN_PERL_VERSION    => '5.010',
    PREREQ_PM           => {            # more to the list $Module::CoreList::version{5.010}
        'Carp'                      => '1.08',
        'Compress::Snappy'          => '0.23',
        'Compress::LZ4Frame'        => '0.012001',
        'Const::Fast'               => '0.014',
        'Data::Compare'             => '1.24',
        'Data::HexDump::Range'      => 0,
        'Data::Validate::Domain'    => '0.11',
        'Data::Validate::IP'        => '0.24',
        'Exception::Class'          => '1.37',
        'Exporter'                  => '5.62',
        'Gzip::Faster'              => '0.19',
        'IO::Select'                => '1.17',
        'List::Util'                => '1.19',
        'Params::Util'              => '1.07',
        'Scalar::Util'              => '1.19',
        'Scalar::Util::Numeric'     => '0.22',
        'Socket'                    => '1.80',
        'String::CRC32'             => '1.5',
        'Sys::SigAction'            => '0.21',
        'Time::HiRes'               => '1.9711',
        'Try::Tiny'                 => '0.21',
    },
    TEST_REQUIRES       => {            # more to the list $Module::CoreList::version{5.010}
        'Capture::Tiny'             => '0.24',
        'Carp'                      => '1.08',
        'Clone'                     => '0.36',
        'Config::IniFiles'          => '2.83',
        'Const::Fast'               => '0.014',
        'Cwd'                       => '3.2501',
        'Data::Dumper'              => '2.121_14',
        'Exporter'                  => '5.62',
        'File::HomeDir'             => '1.00',
        'File::Path'                => '2.04',
        'File::Spec'                => '3.2501',
        'File::Spec::Functions'     => '3.2501',
        'IO::File'                  => '1.14',
        'IO::Socket::INET'          => '1.31',
        'IO::Socket::IP'            => '0.37',
        'List::Util'                => '1.19',
        'Net::EmptyPort'            => 0, # does not have version so any version
        'Params::Util'              => '1.07',
        'Proc::Daemon'              => '0.14',
        'Proc::ProcessTable'        => '0.51',
        'Scalar::Util'              => '1.19',
        'Scalar::Util::Numeric'     => '0.22',
        'Socket'                    => '1.80',
        'Sub::Install'              => '0.927',
        'Sys::SigAction'            => '0.21',
        'Test::Deep'                => '0.119',
        'Test::Exception'           => '0.32',
        'Test::More'                => '0.72',
        'Test::NoWarnings'          => '1.04',
        'Test::TCP'                 => '2.08',
        'Time::HiRes'               => '1.9711',
        'Try::Tiny'                 => '0.21',
    },
    META_MERGE          => {
        no_index                => {
            directory               => [
#                't',
#                'inc',
                'xt',
                'tools',
            ]
        },
        provides                => {
            'Kafka'                 => {
                file                    => 'lib/Kafka.pm',
                version                 => '1.07'
            },
            'Kafka::Connection'     => {
                file                    => 'lib/Kafka/Connection.pm',
                version                 => '1.07'
            },
            'Kafka::Consumer'       => {
                file                    => 'lib/Kafka/Consumer.pm',
                version                 => '1.07'
            },
            'Kafka::Exceptions'     => {
                file                    => 'lib/Kafka/Exceptions.pm',
                version                 => '1.07'
            },
            'Kafka::IO'             => {
                file                    => 'lib/Kafka/IO.pm',
                version                 => '1.07'
            },
            'Kafka::Int64'          => {
                file                    => 'lib/Kafka/Int64.pm',
                version                 => '1.07'
            },
            'Kafka::Internals'      => {
                file                    => 'lib/Kafka/Internals.pm',
                version                 => '1.07'
            },
            'Kafka::Message'        => {
                file                    => 'lib/Kafka/Message.pm',
                version                 => '1.07'
            },
            'Kafka::Producer'       => {
                file                    => 'lib/Kafka/Producer.pm',
                version                 => '1.07'
            },
            'Kafka::Protocol'       => {
                file                    => 'lib/Kafka/Protocol.pm',
                version                 => '1.07'
            },
        },
        'meta-spec'             => { version => 2 },
        resources               => {
            bugtracker              => {
               web                      => 'https://github.com/TrackingSoft/Kafka/issues',
            },
            homepage                => 'https://github.com/TrackingSoft/Kafka',
            repository              => {
                type                    => 'git',
                url                     => 'git://github.com/TrackingSoft/Kafka.git',
                web                     => 'http://github.com/TrackingSoft/Kafka',
            },
        },
    },
    ( $] >= 5.005 ?                             ## Add these new keywords supported since 5.005
        (
            ABSTRACT            => 'Apache Kafka low-level synchronous API, which does not use Zookeeper.',
            AUTHOR              => 'Sergey Gladkov',
        ) : ()
    ),
    dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean               => { FILES => "Debian_CPANTS.txt Kafka-*.tar.gz" },
    LICENSE             => 'perl',
#    NO_MYMETA           => 1,
);

my @patterns = qw( t/*.t );
$config{test} = { TESTS => join ' ', map { glob } @patterns };

if ( $ENV{AUTHOR_TESTS} ) {
    @patterns = qw( xt/author/*.t );
    $config{test}->{TESTS} .= ' '.join ' ', map { glob } @patterns;
}

if ( $ENV{RELEASE_TESTS} ) {
    @patterns = qw( xt/release/*.t );
    $config{test}->{TESTS} .= ' '.join ' ', map { glob } @patterns;
}

mkdir 't/data' unless -d 't/data';
mkdir 'inc' unless -d 'inc';

ExtUtils::MakeMaker::WriteMakefile( %config );
